### DISCUSSION TITLE
How to use Makefile in VScode to create an offline version?

### DISCUSSION THREAD

**Discussion by sk8board** [BIPES/BIPES#225] [Q&A] [answered]
My use of GitHub is with VScode and PlatformIO, so I do not have experience with Makefile for creating an offline version of BIPES. I have searched the internet for how to use Makefile to create an offline version of BIPES, but I have not found a helpful resource.

I have used VScode to clone the master branch of BIPES. Could someone explain how to use Makefile to create the offline version of BIPES?

**Comment by jetannenbaum** [comment]
Hi,

  You need to switch to the offline branch,  I do so on the command line.  git checkout offline.  Once you have the correct branch, you can use cygwin and the cygwinOffline.mak file to make the offline version.  Download cygwin from here https://www.cygwin.com/   cygwin gives you some of the unix tools that will enable you to run the cygwinOffline.mak file.    I'm not familiar with PlatformIO.

   Jim

**Comment by sk8board** [comment, marked as answer]
> You need to switch to the offline branch. Once you have the correct branch, you can use cygwin and the cygwinOffline.mak file to make the offline version. Download cygwin from here https://www.cygwin.com/

Thank you Jim.

For future reference, here are screenshots that shows how to make the offline version with Cygwin.

![image](https://github.com/BIPES/BIPES/assets/96895142/9edeb67c-76ca-48bb-a48e-caf0306ec406)

![image](https://github.com/BIPES/BIPES/assets/96895142/942aaf14-169c-4d6f-8b6a-7120fe1cfe74)

**Reply by jetannenbaum** [reply]
Nicely documented.  If you want, you can edit the cygwinOffline.mak file and change the end of line sequence from CRLF to LF.  That will get rid of the extra character at the end of the indexoffline.html file name.

Jim

**Comment by sk8board** [comment]
Here is additional documentation for how to remove carriage return (CR) from a Windows text file, like `cygwinOffiline.mak`. Removing carriage return (CR) from the `cygwinOffiline.mak` prevents an additional character from being added to the end of the file name, `index_offline.html`, after executing the `cygwinOffiline.mak` file.

1) Reinstall Cygwin, but remember to add the `dos2unix` install package.

2) Before performing the `bash cygwinOffline.mak` command, remove carriage return (CR) from the `cygwinOffiline.mak` file with the following command, `find . -name 'cygwinOffline.mak' |xargs dos2unix`.
